### ISSUE TITLE
Return integers from `charCodeAt` and `codePointAt`

### ISSUE THREAD

**Issue by RReverser** [rustwasm/wasm-bindgen#1362]
Both `String.prototype.charCodeAt` and `String.prototype.codePointAt` are normally expected to return a number.

In case where index is invalid, the return value differs in that the first one would return a `NaN`, whereas the second would return `undefined`, which results in Rust definitions being `f64` and `JsValue` correspondingly.

While these are the nearest types to represent corresponding set of return value, I believe they are not very practical, because getting successful integer result (which you expect most of the time) is cumbersome and requires extra conversions in both cases.

I suggest to align both to return `Option<u32>` instead, which would be already compatible with `code_point_at` out of the box (wasm-bindgen already maps `undefined` to `None` perfectly), and would require a simple conversion for `char_code_at`.

**Comment by alexcrichton** [comment]
Thanks for the report! The `js-sys` wrapper library, though, is intended to be very low level and doesn't have these sorts of abstractions. For that though it'd be perfect for https://github.com/rustwasm/gloo!

Want to move this issue over there for a wrapper crate idea?

**Comment by Pauan** [comment]
@alexcrichton `code_point_at` doesn't need a wrapper though, the type just has to be changed from `JsValue` to `Option<u32>`. That sounds reasonable to me.

I agree that an API that does runtime checking and conversion should probably be in Gloo.

**Comment by alexcrichton** [comment]
Good point! For that I think switching and/or adding a new method to avoid breaking changes sounds like a good idea

**Comment by Pauan** [comment]
I don't really agree with that: the name is already correct, and `Option<u32>` is the correct type, so making up a new name seems really bizarre, and is a learning/discovery hazard.

What I would propose instead is to batch up a bunch of breaking changes, and once we have a sufficient number of them then make a single breaking change release.

That minimizes the amount of times we need to make breaking changes, while still making it possible to do so (to clean up bad APIs, for example).

Of course we can still have a new method in the meantime, but it would just end up being deprecated eventually.

**Comment by Pauan** [comment]
To clarify my position a bit more: I'm definitely not a fan of breaking changes, but I'd be really sad to see a relatively new ecosystem get mired with redundant APIs so soon, for the sake of preserving backwards compatibility at all costs.

I think we still have a little bit of leeway, especially since `code_point_at` is a relatively rarely used API. I do understand the desire for stability, though, especially since Rust + Wasm is gaining a lot of traction. So I'm not going to fight too hard against this.

**Comment by RReverser** [comment]
Yeah I'm fine with keeping `charCodeAt` conversions outside of this, but I think that changing `codePointAt` is reasonable and might be sneaked in as a bugfix (since it should have had such type in the first place)...

**Comment by alexcrichton** [comment]
Oh sorry yeah I should clarify, I definitely agree that if we're doing breaking changes we should just change this outright. If we want to fix this *now* before making a breaking change we'd have to find a new name. In any case the next 0.4 release of js-sys should definitely include this.

I'm not sure if we want to add this now, nor do we really have a schedule for breaking changes. I think it probably wouldn't be the end of the world though to look to doing one soon!

**Comment by RReverser** [comment]
@alexcrichton Actually, if we make this a breaking change, I wonder if we could also build in conversion from JS `Number` to `Option<integers>` that would treat `NaN` as `None`.

There seems to be quite a few precedents in JS where this would help like `charCodeAt` above or `parseInt` etc.

**Comment by Pauan** [comment]
@RReverser `char_code_at` and `parse_int` return `f64`, so the helper methods should go there.

And there already exists an [`is_nan`](https://doc.rust-lang.org/nightly/std/primitive.f64.html#method.is_nan) method for `f64`, so I don't think a helper method would help much.

**Comment by RReverser** [comment]
My point is that 1) these methods return either `NaN` or `u32`/`i32` and 2) casting from `f64` to proper integer type is still a tricky business in Rust since it's way too easy to trigger UB with just `as`, so it would be nice to not put such conversion onto the user.

**Comment by Pauan** [comment]
> casting from `f64` to proper integer type is still a tricky business in Rust since it's way too easy to trigger UB with just `as`

Could you give an example of this? `as` is safe, so it shouldn't be causing any UB.

**Comment by RReverser** [comment]
It's not, unfortunately. See https://github.com/rust-lang/rust/issues/10184 and https://www.reddit.com/r/rust/comments/b1efl5/why_do_these_3_numerical_conversions_f32_as_u8/ for some recent example. 

Exactly because it's not very intuitive / widely known issue, I feel like the conversion should be hidden from user when possible.

**Comment by RReverser** [comment]
To be clear - it's safe for *some* values, and might be fine depending on combination of JS API and target integer type, but that feels pretty error-prone to rely on.

**Comment by RReverser** [comment]
@alexcrichton Either way, do you think we could start a branch for breaking changes for now? It seems there is quite a few issues with this label by now, and it seems worth it to start implementing changes that could be rebase and released all together later.

**Comment by Pauan** [comment]
Oof, that's *really* bad, though it sounds like it'll get fixed in Rust proper? So what's wrong with waiting for the fix?

**Comment by RReverser** [comment]
It will, but it's been around for years, and it seems better to provide an API that avoids foot guns in the first place, especially those as bad as this one.

**Comment by RReverser** [comment]
Another related issue which requires a breaking change: I've just noticed that `from_char_code{1...5}` take `u32` as an argument, but these methods should take `u16` instead.

@alexcrichton Are there any new plans for these breaking changes? Given how they start piling up, it really feels right to at least start a branch with them or something.

**Comment by Pauan** [comment]
> Given how they start piling up, it really feels right to at least start a branch with them or something.

Or at least a document listing all of them, so we don't miss any.

**Comment by RReverser** [comment]
@Pauan I guess `breaking-change` issue label currently serves that purpose relatively well - https://github.com/rustwasm/wasm-bindgen/labels/breaking-change - but it still might take time to actually implement all of them at once. Seems easier to start early and then rebase.

**Comment by alexcrichton** [comment]
@RReverser want to open an issue about u32 -> u16 for these APIs as well?

And yeah, it's intended that the `breaking-change` label is a good bucket for "we'd like to change this in the future"

**Comment by RReverser** [comment]
> want to open an issue about u32 -> u16 for these APIs as well?

Sure, raised https://github.com/rustwasm/wasm-bindgen/issues/1460.
